libraries/standard/mqtt/integration-test/mqtt_system_test.c

```diff
@@ -1312,3 +1382,69 @@ void test_MQTT_Restore_Session_Duplicate_Incoming_Publish_Qos2( void )
     /* Make sure that the library cleared the record for the incoming QoS 2 PUBLISH packet. */
     TEST_ASSERT_EQUAL( MQTT_PACKET_ID_INVALID, context.incomingPublishRecords[ 0 ].packetId );
 }
+
+/**
+ * @brief Verifies that the library supports notifying the broker to retain a PUBLISH message
+ * for a topic using the retain flag.
+ */
+void test_MQTT_Publish_With_Retain_Flag( void )
+{
+    /* Publish to a topic with the "retain" flag set. */
+    TEST_ASSERT_EQUAL( MQTTSuccess, publishToTopic( &context,
+                                                    TEST_MQTT_TOPIC,
+                                                    true,  /* setRetainFlag */
+                                                    false, /* isDuplicate */
+                                                    MQTTQoS1,
+                                                    MQTT_GetPacketId( &context ) ) );
+    /* Complete the QoS 1 PUBLISH operation. */
+    TEST_ASSERT_FALSE( receivedPubAck );
+    TEST_ASSERT_EQUAL( MQTTSuccess,
+                       MQTT_ProcessLoop( &context, 2 * MQTT_PROCESS_LOOP_TIMEOUT_MS ) );
+    TEST_ASSERT_TRUE( receivedPubAck );
+
+    /* Subscribe to the same topic that we published the message to.
+     * The broker should send the "retained" message with the "retain" flag set. */
+    TEST_ASSERT_EQUAL( MQTTSuccess, subscribeToTopic(
+                           &context, TEST_MQTT_TOPIC, MQTTQoS1 ) );
+    TEST_ASSERT_FALSE( receivedSubAck );
+    TEST_ASSERT_EQUAL( MQTTSuccess,
+                       MQTT_ProcessLoop( &context, 2 * MQTT_PROCESS_LOOP_TIMEOUT_MS ) );
+    TEST_ASSERT_TRUE( receivedSubAck );
```

Comment:
I see this pattern a lot in this code.  We don't have any guarantees about broker latency, do we?

Reply:
I would have to look into the Mosquitto configuration options to check whether there is any latency guarantee we can get. But yes, if we don;t have, then these tests are dependent on the responsiveness of the broker..

Reply:
We discussed offline, and Dan mentioned that he was concerned about the timeout value introducing flaky behavior of these network-dependent integration tests. 
As these tests are run on hosts (or instances) with ethernet connection, a flaky behavior of these tests is less likely (based on the runs of the `v4_beta` branch that have been fairly stable).  Anyway, we will get concrete data once we have enabled nightly CI runs for the demos. We don't expect a flaky behavior when running against a locally owned Docker container running the Mosquitto broker